skrifa/src/patchmap.rs

```diff
@@ -3,223 +3,295 @@
+            let index = i + cumulative_entry_map_count;
+            let byte_index = (index * field_width * 2) as usize;
+            let data = FontData::new(&feature_map.entry_map_data()[byte_index..]);
+            let mapped_entry_index = record.first_new_entry_index().get() + i;
+            let record = EntryMapRecord::read(data, max_entry_index)?;
+            let first = record.first_entry_index().get();
+            let last = record.last_entry_index().get();
+            if first > last
+                || first > max_glyph_map_entry_index
+                || last > max_glyph_map_entry_index
+                || mapped_entry_index <= max_glyph_map_entry_index
+                || mapped_entry_index > max_entry_index
+            {
+                // Invalid, continue on
+                continue;
+            }
 
-    /// Produce a mapping from each glyph id to the codepoint(s) that map to that glyph.
-    fn glyph_to_unicode_map<'a>(font: &impl TableProvider<'a>) -> HashMap<GlyphId, IntSet<u32>> {
-        let charmap = Charmap::new(font);
-        let mut glyph_to_unicode = HashMap::<GlyphId, IntSet<u32>>::new();
-        for (cp, glyph) in charmap.mappings() {
-            glyph_to_unicode
-                .entry(glyph)
-                .or_insert_with(IntSet::<u32>::empty)
-                .insert(cp);
+            if entries.intersects_range(first..=last) {
+                entries.insert(mapped_entry_index);
+            }
         }
-        glyph_to_unicode
+        next_tag = tag_it.next();
     }
 
-    fn apply_uri_template(uri_template: &str, _entry_index: usize) -> String {
-        // TODO(garretrieger): properly implement this, may deserve to go into it's own module.
-        uri_template.to_string()
-    }
+    Ok(())
+}
 
-    // TODO(garretrieger): add template application method that takes a string entry id (will be needed for format 2)
+#[derive(Clone, Eq, PartialEq, Debug, Hash, Copy)]
+pub enum PatchEncoding {
+    Brotli,
+    PerTableBrotli { fully_invalidating: bool },
+    GlyphKeyed,
 }
 
-#[derive(Clone, Eq, PartialEq, Hash)]
-pub struct PatchUri {
-    uri: String,
-    encoding: PatchEncoding,
+impl PatchEncoding {
+    fn from_format_number(format: u8) -> Option<Self> {
+        // Based on https://w3c.github.io/IFT/Overview.html#font-patch-formats-summary
+        match format {
+            1 => Some(Self::Brotli),
+            2 => Some(Self::PerTableBrotli {
+                fully_invalidating: true,
+            }),
+            3 => Some(Self::PerTableBrotli {
+                fully_invalidating: false,
+            }),
+            4 => Some(Self::GlyphKeyed),
+            _ => None,
+        }
+    }
 }
 
-#[derive(Clone, Eq, PartialEq, Hash)]
-pub struct Entry {
-    pub patch_uri: PatchUri,
-    pub codepoints: IntSet<u32>,
-    pub feature_tags: BTreeSet<Tag>,
-    pub compatibility_id: [u32; 4],
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct PatchUri {
```

Comment:
Same here wrt doc comment. Template seems to imply that the string contains a pattern that is replaced with the index?

Reply:
That's right, added a comment and linked to the relevant parts of the spec.